### PR TITLE
Implement and raise Kernel typespecerror

### LINKED ISSUE #12277: Do not raise compile error from Kernel.Typespec
Instead we should raise `Kernel.TypespecError` that works exactly the same as `CompileError`. The raising code is already defined and encapsulated here:

https://github.com/elixir-lang/elixir/blob/main/lib/elixir/lib/kernel/typespec.ex#L942-L944

### PR DESCRIPTION
**Pull request by zacck**
fixes #12277 

Chore to implement and raise the Kernel.TypespecError

### CODE REVIEW

File: lib/elixir/lib/exception.ex [hunk 1]
```diff
@@ -955,6 +955,18 @@ defmodule CompileError do
   end
 end
 
+defmodule Kernel.TypespecError do
+  defexception [:file, :line, description: "typespec error"]
```
Comment (whatyouhide) [inline lib/elixir/lib/exception.ex:959]: Do we need a default `description` here?

Comment (zacck) [reply lib/elixir/lib/exception.ex:959]: I bet you one wouldn't hurt, let me add one

Comment (zacck) [reply lib/elixir/lib/exception.ex:959]: @whatyouhide  I am not very sure we do, I am still trying to figure out why we had to change to this perhaps that would give me more context

Comment (josevalim) [reply lib/elixir/lib/exception.ex:959]: We always give a description, so we can remove this one:

```suggestion
  defexception [:file, :line, :description]
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**